### PR TITLE
Move Configuration Provider from Kernel

### LINKED ISSUE #1: Move Configuration Provider from Kernel
In Carbon 5.2.0-m3, configuration provider and org.wso2.carbon.extensions.configuration.maven.plugin resides in kernel and supports only OSGi model. This needs to be moved to this repo and in-addition to OSGi support, it should also provide support for non-OSGi. 

deployment.yaml should be runtime specific, i.e., \<CARBON HOME\>/wso2/\<runtime\>/config

### PR DESCRIPTION
**Pull request by vidurananayakkara**
Resolves #1

### CODE REVIEW

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    /**
+     * Initialise carbon config provider.
+     */
+    private void initializeConfigProvider() {
+        if (!ConfigProviderDataHolder.getInstance().getSecureVault().isPresent() ||
+                !ConfigProviderDataHolder.getInstance().getBundleContext().isPresent()) {
+            logger.debug("Waiting for config provider dependencies");
+            return;
+        }
```
Comment (daneshk) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: Do we need this. in osgi mode. we wait till the reference component active.

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        // TODO: Get carbon config location from carbon utils
+        // TODO: Support getting configProviderImpl in non-OSGi mode
+        Path deploymentConfigPath =
+                Paths.get(System.getProperty(Constants.CARBON_HOME), "conf", Constants.DEPLOYMENT_CONFIG_YAML);
+        ConfigFileReader configFileReader = new YAMLBasedConfigFileReader(deploymentConfigPath);
+        ConfigProvider configProvider = new ConfigProviderImpl(configFileReader);
+        ConfigProviderDataHolder.getInstance().getBundleContext()
+                .ifPresent(bundleContext -> bundleContext
+                        .registerService(ConfigProvider.class, configProvider, null));
```
Comment (daneshk) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: bundle context is always present. no need to check it again.

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    @Reference(
+            name = "org.wso2.carbon.securevault",
+            service = SecureVault.class,
+            cardinality = ReferenceCardinality.MANDATORY,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unRegisterSecureVault"
+    )
+    protected void registerSecureVault(SecureVault secureVault) {
+        ConfigProviderDataHolder.getInstance().setSecureVault(secureVault);
+        initializeConfigProvider();
```
Comment (daneshk) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: why are we calling initializaConfigProvider here?

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        immediate = true
+)
+public class ConfigProviderComponent {
+    private static final Logger logger = LoggerFactory.getLogger(ConfigProviderComponent.class);
+
+    @Activate
+    protected void activate(BundleContext bundleContext) {
+        ConfigProviderDataHolder.getInstance().setBundleContext(bundleContext);
+        initializeConfigProvider();
+        logger.debug("Activating Carbon Configuration Component");
```
Comment (daneshk) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: Can you change the debug message to something like ""Carbon Configuration Component activated"". Component is already activated at this point.

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/reader/ConfigFileReader.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            String message = "Error while reading the configuration file, file path is null";
+            LOGGER.error(message);
+            throw new ConfigurationException(message);
+        }
+        try {
+            byte[] contentBytes = Files.readAllBytes(configurationFilePath);
+            return new String(contentBytes, StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            String message = "Error while reading configuration file";
+            LOGGER.error(message);
```
Comment (daneshk) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/reader/ConfigFileReader.java]: Need to log the exception stacktrace here.

File: components/org.wso2.carbon.configuration/src/test/resources/conf/deployment.properties [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+[Example.properties]/transport.abc.secure=${sys:abc.port.secure,true}
+#[Example.properties]/transport.xyz.port=9090
+#[Example.properties]/transport.xyz.secure=true
+
+#Invalid entries
+[Example3.xml]/configurations/tenant=${sys:sample.tenant}
+[Example4.xml]/configurations/transports/transport/port=${env:sample.port}
+
+[Example5.xml]/configurations/transports/transport[name\='abc']/@secure=${env:sample.abc.port,8081}
+[Example6.xml]/configurations/transports/transport[name\='xyz']/@secure=${sys:sample.xyz.port,9091}
```
Comment (daneshk) [inline components/org.wso2.carbon.configuration/src/test/resources/conf/deployment.properties]: Please check and remove unnecessary test resource files. This file not use.

File: tests/org.wso2.carbon.configuration.test/src/test/resources/META-INF/services/javax.annotation.processing.Processor [hunk 7]
```diff
@@ -0,0 +1 @@
+org.wso2.carbon.configuration.ConfigurationProcessor
```
Comment (daneshk) [inline tests/org.wso2.carbon.configuration.test/src/test/resources/META-INF/services/javax.annotation.processing.Processor]: we don't need to add service provider for the testcase. It should come from the configuration dependency

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java (hunk 1 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: Yes, Carbon config is dependant on Secure Vault. Therefore we only need to register the configuration service once the secure vault service is registered as an OSGi service

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java (hunk 2 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: We set the bundle context in the `@Activate` method in the component. Therefore when this method is called bundle context may not be present

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java (hunk 3 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: When the SecureVault is registered we need to initialize the config provider (if it is possible only)

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java (hunk 4 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: Changed as requested

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/reader/ConfigFileReader.java (hunk 5 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/reader/ConfigFileReader.java]: Exception stack trace will be logged

File: components/org.wso2.carbon.configuration/src/test/resources/conf/deployment.properties (hunk 6 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.configuration/src/test/resources/conf/deployment.properties]: Unnecessary test resources are removed

File: tests/org.wso2.carbon.configuration.test/src/test/resources/META-INF/services/javax.annotation.processing.Processor (hunk 7 above)
Comment (vidurananayakkara) [inline tests/org.wso2.carbon.configuration.test/src/test/resources/META-INF/services/javax.annotation.processing.Processor]: Changed as requested

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java (hunk 1 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: Changed as requested

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java (hunk 2 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: Changed as requested

File: components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java (hunk 3 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java]: Changed as requested

**Comment by vidurananayakkara** [comment]
Closing this PR since opening a new PR with the proper git commit history included. Please note that all the code review suggestions are applied in the new PR

**Pull request closed without merging**